turbopack/crates/turbo-persistence/src/compaction/interval_map.rs

```diff
@@ -0,0 +1,166 @@
+                } else {
+                    None
+                };
+                self.intervals.insert(
+                    index,
+                    IntervalPoint {
+                        start: location,
+                        value,
+                    },
+                );
+                index
+            }
+        }
+    }
+
+    /// Applies the update function to all values in the specified range.
+    pub fn update(&mut self, range: &(u64, u64), mut update: impl FnMut(&mut T))
+    where
+        T: Default + Clone,
+    {
+        let start = range.0;
+        let end = range.1;
+        if start > end {
+            return;
+        }
+
+        let start = self.ensure_point(start);
+        if end == u64::MAX {
+            for i in start..self.intervals.len() {
+                update(self.intervals[i].value.get_or_insert_default());
+            }
+        } else {
+            let end = self.ensure_point(end + 1);
+
+            for i in start..end {
+                update(self.intervals[i].value.get_or_insert_default());
+            }
+        }
+    }
+
+    /// Tests if any values in the specified range satisfy the predicate.
+    pub fn test(&self, range: &(u64, u64), mut predicate: impl FnMut(&T) -> bool) -> bool {
+        let start = range.0;
+        let end = range.1;
+        if start > end {
+            return false;
+        }
+
+        let start_index = self
+            .intervals
+            .binary_search_by_key(&start, |point| point.start)
+            .unwrap_or_else(|x| x);
+
+        let end_index = match self
+            .intervals
+            .binary_search_by_key(&end, |point| point.start)
+        {
+            Ok(index) => index + 1,
+            Err(index) => index,
+        };
+
+        for i in start_index..end_index {
+            if let Some(value) = &self.intervals[i].value
+                && predicate(value)
+            {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Returns an iterator over the ranges and their associated values.
+    pub fn ranges(&self) -> impl Iterator<Item = ((u64, u64), &T)> {
+        (0..self.intervals.len()).filter_map(move |i| {
+            let start = self.intervals[i].start;
+            let end = if i + 1 < self.intervals.len() {
+                self.intervals[i + 1].start - 1
+            } else {
+                u64::MAX
+            };
+            self.intervals[i]
+                .value
+                .as_ref()
+                .map(|value| ((start, end), value))
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_interval_map() {
+        let mut map = IntervalMap::new();
+        map.update(&(5, 15), |v| *v += 1);
+        map.update(&(10, 15), |v| *v += 2);
+        map.update(&(10, 20), |v| *v += 4);
+        map.update(&(0, u64::MAX), |v| *v += 8);
+        map.update(&(15, 20), |v| *v += 16);
```

Comment:
Nit: Use bitwise OR here, similar to your `expected` below so that it's more obvious what's happening.

```suggestion
        // set bitflags for these ranges
        map.update(&(5, 15), |v| *v |= 1);
        map.update(&(10, 15), |v| *v |= 2);
        map.update(&(10, 20), |v| *v |= 4);
        map.update(&(0, u64::MAX), |v| *v |= 8);
        map.update(&(15, 20), |v| *v |= 16);
```